### PR TITLE
Fix issues with HTMLForm::MultipartReadBuffer and PeekableReadBuffer

### PR DESCRIPTION
**Pull request by abyss7**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix reading the HTTP POST request with "multipart/form-data" content type

### CODE REVIEW

File: src/IO/PeekableReadBuffer.cpp [hunk 1]
```diff
@@ -134,10 +143,11 @@ bool PeekableReadBuffer::nextImpl()
 
     checkStateCorrect();
     bool res;
+    bool checkpoint_at_end = checkpoint && *checkpoint == working_buffer.end();
 
     if (checkpoint)
     {
-        if (currentlyReadFromOwnMemory())
+        if (currentlyReadFromOwnMemory() || checkpoint_at_end)
```
Comment (tavplubix) [inline src/IO/PeekableReadBuffer.cpp:150]: What is the purpose if this change? Seems like the case when checkpoint is at the end of working buffer is handled [here](https://github.com/ClickHouse/ClickHouse/blob/414f470c79eb22b0ca47b82f11625cf80b0231aa/src/IO/PeekableReadBuffer.cpp#L56).

Comment (abyss7) [reply src/IO/PeekableReadBuffer.cpp:150]: It's for the case when we put checkpoint on the edge of internal buffer and then call `next()`.

Comment (abyss7) [reply src/IO/PeekableReadBuffer.cpp]: BTW, it's really redundant here - removed.

File: src/IO/PeekableReadBuffer.cpp [hunk 2]
```diff
@@ -113,12 +113,21 @@ void PeekableReadBuffer::rollbackToCheckpoint(bool drop)
 {
     checkStateCorrect();
 
-    if (!checkpoint)
-        throw DB::Exception("There is no checkpoint", ErrorCodes::LOGICAL_ERROR);
```
Comment (tavplubix) [inline src/IO/PeekableReadBuffer.cpp:117]: Ok. We can just remove [the check](https://github.com/ClickHouse/ClickHouse/blob/27600064159ac63e05800ef37afa41b317771352/src/IO/tests/gtest_peekable_read_buffer.cpp#L51-L64) for this.

**Review by tavplubix — APPROVED** [review]

**Pull request merged**